fendermint/vm/topdown/proof-service/src/f3_client.rs

```diff
@@ -0,0 +1,259 @@
+        power_table: filecoin_f3_gpbft::PowerEntries,
+    ) -> Result<Self> {
+        let light_client = LightClient::new(rpc_endpoint, network_name)
+            .context("Failed to create F3 light client")?;
+
+        // Initialize state with provided power table from actor
+        let state = LightClientState {
+            instance: initial_instance,
+            chain: None,
+            power_table,
+        };
+
+        info!(
+            initial_instance,
+            power_table_size = state.power_table.len(),
+            network = network_name,
+            rpc = rpc_endpoint,
+            "Created F3 client with power table from F3CertManager actor"
+        );
+
+        Ok(Self {
+            light_client: Arc::new(Mutex::new(light_client)),
+            state: Arc::new(Mutex::new(state)),
+        })
+    }
+
+    /// Create F3 client by fetching power table from RPC (TESTING ONLY)
+    ///
+    /// For testing/development. In production, use `new()` with power table from
+    /// the F3CertManager actor on-chain.
+    ///
+    /// # Arguments
+    /// * `rpc_endpoint` - F3 RPC endpoint
+    /// * `network_name` - Network name (e.g., "calibrationnet", "mainnet")
+    /// * `initial_instance` - F3 instance to bootstrap from
+    pub async fn new_from_rpc(
+        rpc_endpoint: &str,
+        network_name: &str,
+        initial_instance: u64,
+    ) -> Result<Self> {
+        let mut light_client = LightClient::new(rpc_endpoint, network_name)
+            .context("Failed to create F3 light client")?;
+
+        // Fetch initial power table from RPC (for testing)
+        let state = light_client
+            .initialize(initial_instance)
+            .await
+            .context("Failed to initialize light client with power table from RPC")?;
+
+        info!(
+            initial_instance,
+            power_table_size = state.power_table.len(),
+            network = network_name,
+            "Created F3 client with power table from RPC (testing mode)"
+        );
+
+        Ok(Self {
+            light_client: Arc::new(Mutex::new(light_client)),
+            state: Arc::new(Mutex::new(state)),
+        })
+    }
+
+    /// Fetch and validate an F3 certificate
+    ///
+    /// This performs full cryptographic validation including:
+    /// - BLS signature correctness
+    /// - Quorum requirements (>2/3 power)
+    /// - Chain continuity (sequential instances)
+    /// - Power table validity
+    ///
+    /// # Returns
+    /// `ValidatedCertificate` containing the cryptographically verified certificate
+    pub async fn fetch_and_validate(&self, instance: u64) -> Result<ValidatedCertificate> {
+        debug!(instance, "Starting F3 certificate fetch and validation");
+
+        // STEP 1: FETCH certificate from F3 RPC
+        let fetch_start = Instant::now();
+        let f3_cert = match self
+            .light_client
+            .lock()
+            .await
+            .get_certificate(instance)
```

Comment:
This will generate proofs for instances even if passed/expired long time ago. If the purpose of this cache service is to be used by block proposers, only latest instances which were not checkpointed previously are relevant.